src/transformers/vocabulary2id.py

```diff
@@ -0,0 +1,71 @@
+import operator
+
+from pyspark import RDD, Row
+from models.code2vec import Code2Vec
+
+from ast import literal_eval as make_tuple
+from sourced.ml.transformers import Transformer
+
+
+class Vocabulary2Id(Transformer):
+    def __init__(self, keep_freq: bool, output: str, **kwargs):
+        super().__init__(**kwargs)
+        self.keep_freq = keep_freq
+        self.output = output
+
+    def __call__(self, rows: RDD):
+        values, paths = self.build_vocabularies(rows)
+
+        doc2path_contexts = self.build_doc2pc(rows)
+
+        values = values.collect()
+        paths = paths.collect()
+
+        doc2path_contexts = doc2path_contexts.collect()
+
+        Code2Vec().construct(values=values, paths=paths, path_contexts=doc2path_contexts).save(
+            self.output)
+
+    @staticmethod
+    def _unstringify_path_context(row):
+        """
+        Takes a row containing ((pc, doc), freq) and returns a tuple (removes namespace prefix v.)
+        """
+        return make_tuple(row[0][0][2:])
+
+    def _get_path(self, row: Row):
+        # 2: removes the namespace v. from the string to parse it as tuple
+        k, v = Vocabulary2Id._unstringify_path_context(row), row[1]
+        return [(k[0], 1), (k[1], 1), (k[2], 1)]
+
+    def build_vocabularies(self, rows: RDD):
+        """
+        Process rows to gather values and paths with/without their frequencies.
+        """
+
+        r = rows \
```

Comment:
rename r to rows and cache after reduceByKey or you will do it twice, which is time consuming, dont forget to uncache after both collects

Reply:
not sure how things are cached / uncached in Spark I'm pretty new to it, could you give more details?

Reply:
sure. in spark you can either do transformations or actions on RDDs/Dataframes/etc. Those objects are lazily evaluated, which means transformations are not applied if an action does not trigger them: 
```
r.map(lambda x: x[0]).filter(my_condition)  # does nothing 
r.map(lambda x: x[0]).filter(my_condition).collect()  # the action collect triggers transformations  
```
in order to not repeat parts of the pipeline, you can cache the data in RDD or on disk, here:
```
rows = rows \
    .flatMap(self._get_path) \
    .reduceByKey(operator.add) \
    .persist()  # nothing happens

values = rows.filter(lambda x: type(get_elem(x)) == str).collect() 
# after this, rows is cached, so doing the second collect starts from 
# after the reduceByKey, not before the flatmap
paths = r.filter(lambda x: type(get_elem(x)) == tuple).collect()
rows.unpersist(). # or this will stay in memory/disk
```

we usually add a persist argument to select level of persistence, for more see [here](https://github.com/src-d/ml/blob/master/doc/spark.md)